add blinking to species

qrwas opened the pull request:
## About the PR
Adds blinking to all species and mouses, and new trait blink dyspraxia

## Why / Balance
For some reason, I thought characters already blinked and could close their eyes. But as it turned out, they didn't. Outraged by this, I decided to make this PR to add some immersion. It makes our characters blink and close their eyes when they sleep!

## Technical details
This adds a new "EyeBlinkingSystem" that:
- listens for an emotion event on the server to force an entity to blink
- makes entities blink on the client side only
- A new component is added to entities or eyes that should blink, specifying the path to eyelid sprites and other parameters like blink speed, asynchronous blinking, etc.

## Media
 Blinks: (old)
https://github.com/user-attachments/assets/92e85eec-f7aa-4bea-81cf-b67589008aaf

Blink and close eyes:
https://github.com/user-attachments/assets/f5575dd9-622e-4cb5-9472-4f1b5f59f84c

blink dyspraxia:

https://github.com/user-attachments/assets/aac86fe6-837c-4362-a2d0-704c4aaf1966

Changeling:
https://cdn.discordapp.com/attachments/1524193921872433222/1535313349053653105/0807.mp4?ex=6a774f7b&is=6a75fdfb&hm=c4e7542e4ba588d662ed95a27f08ccbede9249122a35840ffb33a32405e2e6c2&

All species:
https://discord.com/channels/310555209753690112/1524193921872433222/1535316815582990456

## Test Plan
- Spawn all species
- Wait and check that they are all blinking
- Disable character randomization via cvar (since traits don't work for spawned characters)
- Set the new Blink Dyspraxia trait in the character editor
- Join the game, wait for the character to blink, and verify that their eyes are desynchronized
- Try to blink using the @blink emote — the character should blink, and if they have the trait, their eyes should still be desynchronized
- Close eyes via the action bar — the character's eyes should close, and their sight/vision should be restricted
- Pick up a welding tool, try welding a door, then open eyes — they should be fine (no post-welding flash/burn effects)
- Try to blink using the emote — the character should blink
- Go to sleep, try to open eyes via the action — they shouldn't open
- Wake up, try closing and opening eyes, try blinking
- Take a changeling, absorb another character's DNA, transform — eyes should return to normal

## Requirements
- [x] I have read and am following the [Pull Request and Changelog Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html).
- [x] I have added media to this PR or it does not require an in-game showcase.

## Breaking Changes
- A new layer, HumanoidVisualLayers.Eyelids, has been added to all species after Eyes. This is required for the EyeBlinking component
- A new layer HumanoidVisualLayers.OverEyes has  been added to all species after Eyelids. This required for gauzes
- All eye gauzes now applied to OverEyes
- Removed Content.Shared\Eye\Blinding\Components\EyeClosingComponent.cs and Content.Shared\Eye\Blinding\Systems\EyeClosingSystem.cs. The eye closing action has been moved to the new system.


**Changelog**
:cl:
- add: Species can now blink!
- add: Species now close their eyes when sleeping!
- add: New trait "blink dyspraxia"
- add: New action to close eyes
- tweak: When the emotion "blink", your character will blink!

qrwas committed:
add blinking

qrwas committed:
Rename Enum

beck-thompson commented on Content.Shared/Eye/Blinking/EyeBlinkingComponent.cs:
```diff
@@ -0,0 +1,32 @@
+
+namespace Content.Shared.Eye.Blinking;
+[RegisterComponent, NetworkedComponent]
+[AutoGenerateComponentState, AutoGenerateComponentPause]
```
make sure to add a comment on the component!

qrwas commented on Content.Shared/Eye/Blinking/SharedEyeBlinkingSystem.cs:
```diff
@@ -0,0 +1,116 @@
+        OpenEyes(ent.Owner, ent.Comp);
+    }
+
+    private void OnComponentRemoved(RemovedComponentEventArgs args)
```
Will HasComp be more productive?

qrwas committed:
Some refactor to check Sleeping and add comments

Lukasz825700516 commented on Content.Shared/Eye/Blinking/SharedEyeBlinkingSystem.cs:
```diff
@@ -0,0 +1,105 @@
+            else
+            {
+                if (comp.NextOpenEyesTime <= curTime)
+                    OpenEyes(uid, comp);
```
Should it not use the same method as cameleon clothing scrambling animation, to not flood the server with component getting Dirtied every few seconds?

Lukasz825700516 commented on Content.Shared/Eye/Blinking/SharedEyeBlinkingSystem.cs:
```diff
@@ -0,0 +1,105 @@
+    override public void Initialize()
+    {
+        base.Initialize();
+        SubscribeLocalEvent<EyeBlinkingComponent, SleepStateChangedEvent>(SleepStateChangedEventHanlder);
```
Shouldnt it be more subscribtion to BlindnessChangeEvent, to also stop blinking when you close eyes, or have tear gas in them?

slarticodefast commented on Content.Shared/Eye/Blinking/EyeBlinkingComponent.cs:
```diff
@@ -0,0 +1,36 @@
+    public TimeSpan NextBlinkingTime;
+    // Whether the entity is currently sleeping.
+    [DataField, AutoNetworkedField]
+    public bool IsSleeping;
```
this information is already part of other components, no need to duplicate it here

slarticodefast reviewed (changes requested):
This introduces a lot of networking for mobs for just a few pixels that are changed.
This should probably be completely client side, no need for this to have a synced authorative server state. Or even simpler, you can just make separate animated sprites for the eye layer and change that via the appearance data and generic visualizer. The blinking animation will then automatically happen without having to manually close and open eyes.

Finally this has some gameplay implications: You will no longer be able to fake your death by dropping down and using the death gasp emote since you will still be blinking.

Huaqas commented:
Does this work with heterochromia markings?

FairlySadPanda reviewed (commented):
Grumble: I think what this ends up being is part of an idle animation system, of which eye blinking is one layer. Adding in a component to do just this rather than doing the engineering task to create a mob animation system is the wrong approach.

That said, nobody has made such a system, I am not working on such a system, and this is something players will love, so I think it's sensible within the confines of what a contributor can get committed in this domain of work at the moment.

qrwas committed:
Refactoring EyeBlinking

qrwas committed:
Update summary and remove old code

qrwas committed:
Another refactoring

qrwas committed:
pupupu....

qrwas committed:
Refactoring blink. again

qrwas committed:
Update EyeBlinkingSystem.cs

qrwas committed:
Another update Blinking

qrwas committed:
Update SharedEyeBlinkingSystem.cs

qrwas committed:
cleanup

qrwas commented:
I need help and advice. Is the blinking approach I chose correct?
I made the blink emote but it stops working after the character falls asleep and wakes up, although the automatic blinking continues...

qrwas commented:
The only thing left for me now is to make them blink after sleep ,_,
I just don't understand which way to think and where to look for the mistake.

qrwas committed:
another try to fix sleeping

qrwas committed:
refactoring to use NetEvent. I dumb for apperance ._.

qrwas committed:
Update blinking_emotes.yml

qrwas committed:
Add new layer for fix eye marking

qrwas committed:
Fix state

qrwas committed:
Add EyelidsColor for component

qrwas commented:
> Does this work with heterochromia markings?

I finally solved the problem with marking for blinking
now you can do heterochromia and it will work fine with the eyelids. 
So yes. Now it works with heterochromia markings!

qrwas committed:
Update HumanoidVisualLayersExtension.cs

qrwas committed:
Small refactoring and add mouse blinking

beck-thompson commented on Content.Client/Eye/Blinking/EyeBlinkingSystem.cs:
```diff
@@ -0,0 +1,113 @@
+
+        if (ent.Comp.EyesClosed)
+            return;
+
```
remove space

beck-thompson commented on Content.Shared/Eye/Blinking/SharedEyeBlinkingSystem.cs:
```diff
@@ -0,0 +1,103 @@
+public enum EyeBlinkingVisuals : byte
+{
+    EyesClosed
+}
```
newline after

beck-thompson commented on Content.Client/Eye/Blinking/EyeBlinkingSystem.cs:
```diff
@@ -0,0 +1,113 @@
+        SubscribeLocalEvent<EyeBlinkingComponent, ComponentStartup>(OnStartup);
+    }
+
+    private void OnStartup(Entity<EyeBlinkingComponent> ent, ref ComponentStartup args)
```
Is there any way to get rid of this? It suses me out big time.

beck-thompson commented on Content.Shared/Eye/Blinking/SharedEyeBlinkingSystem.cs:
```diff
@@ -0,0 +1,103 @@
+    public virtual void BlindnessChangedEventHanlder(Entity<EyeBlinkingComponent> ent, ref BlindnessChangedEvent args)
+    {
+        ent.Comp.EyesClosed = args.Blind;
+        var ev = new ChangeEyeStateEvent(GetNetEntity(ent.Owner), args.Blind);
```
Why can't you raise the event on the entity itself?

beck-thompson reviewed (changes requested):
A few things I noticed! Looks pretty good though 👍

slarticodefast commented on Content.Client/Eye/Blinking/EyeBlinkingSystem.cs:
```diff
@@ -0,0 +1,113 @@
+    {
+        var ent = GetEntity(ev.NetEntity);
+
+        if (!ent.IsValid() || !TryComp<EyeBlinkingComponent>(ent, out var blinkingComp))
```
why is this not a directed event?

slarticodefast commented on Content.Shared/Eye/Blinking/SharedEyeBlinkingSystem.cs:
```diff
@@ -0,0 +1,103 @@
+    protected virtual void ResetBlink(Entity<EyeBlinkingComponent> ent)
+    {
+        ent.Comp.NextBlinkingTime = _timing.CurTime + ent.Comp.BlinkInterval + ent.Comp.BlinkDuration;
+        Dirty(ent);
```
The blinking logic should just be a client side animation instead of networking it every single update for every single mob

qrwas committed:
Start refactoring to client side

qrwas committed:
Another refactoring blinking

qrwas committed:
Update EyeBlinkingComponent.cs

qrwas committed:
Some rename and update comments

qrwas committed:
Update EyeBlinkingSystem.cs

qrwas commented:
Added comments, and moved the main blinking logic to the client. About using network events, I hope, I answered. If you know a way to get around this - I will only agree to improve this code, and accordingly my skills, to avoid such mistakes in the future

beck-thompson commented on Content.Client/Eye/Blinking/EyeBlinkingSystem.cs:
```diff
@@ -0,0 +1,162 @@
+    }
+
+
+    private void OnStartup(Entity<EyeBlinkingComponent> ent, ref ComponentStartup args)
```
use map init. Otherwise the urst blink in the spawn menu...

beck-thompson commented on Content.Client/Eye/Blinking/EyeBlinkingSystem.cs:
```diff
@@ -0,0 +1,162 @@
+    {
+        var minInterval = ent.Comp.MinBlinkInterval;
+        var maxInterval = ent.Comp.MaxBlinkInterval;
+        var randomSeconds = minInterval + (_random.NextDouble() * (maxInterval - minInterval));
```
dont copy paste the random interval code, make a function and reuse it 🤔

beck-thompson commented on Content.Client/Eye/Blinking/EyeBlinkingSystem.cs:
```diff
@@ -0,0 +1,162 @@
+        if (!ent.Owner.IsValid())
+            return;
+
+        ResetBlink(ent);
```
Probably put this after the eyes closed and blink in progress

beck-thompson commented on Content.Shared/Eye/Blinking/EyeBlinkingComponent.cs:
```diff
@@ -0,0 +1,88 @@
+    /// Indicates whether a blink is currently in progress.
+    /// </summary>
+    [DataField, AutoNetworkedField]
+    public bool BlinkInProgress = false;
```
What is the difference  between blink in progress and eyes closed?

beck-thompson commented on Resources/Textures/Interface/Emotes/attributions.yml:
```diff
@@ -130,3 +130,8 @@
+- files: ["blink.png"]
+  license: "CC-BY-SA-3.0"
+  copyright: "Original sprite 'laugh' created by Sarahon, modified by qrwas"
+  source: "https://github.com/qrwas"
```
add newline at the end

beck-thompson commented on Content.Client/Eye/Blinking/EyeBlinkingSystem.cs:
```diff
@@ -0,0 +1,162 @@
+        _sprite.LayerSetColor(layer, eyeClsoed ? blinkColor : Color.Transparent);
+    }
+
+    public void Blink(Entity<EyeBlinkingComponent> ent)
```
add a comment if its public! All public functions should have comments

qrwas commented on Content.Shared/Eye/Blinking/EyeBlinkingComponent.cs:
```diff
@@ -0,0 +1,88 @@
+    /// The prototype ID of the emote that triggers a forced blink.
+    /// </summary>
+    [DataField, AutoNetworkedField]
+    public ProtoId<EmotePrototype> BlinkEmoteId = "Blink";
```
I think the necessary trigger words are defined in the emotion prototype

qrwas committed:
some fixes and add comments

qrwas committed:
Update meta.json

qrwas committed:
Refactoring to use apperanceSystem

qrwas committed:
add some check if comp enabled

qrwas committed:
pupupu

qrwas committed:
Try to fix working blinking on changeling

qrwas committed:
Some another fixes eyelids sprite with changeling

qrwas committed:
Some fixes

qrwas committed:
Update attributions.yml

qrwas commented:
in the process of implementing additional complications and new blinking parameters

qrwas committed:
Start refactoring to more eyelids

qrwas committed:
Remade to use multiple layers

qrwas committed:
Merge branch 'master' into Add-eyes-blinking

qrwas committed:
Update human blinking

qrwas committed:
Add arachnid blinking

qrwas committed:
Fix update upperance

qrwas committed:
Delete eyes.png

qrwas committed:
add new eyelids for other species

qrwas committed:
another fix naming

qrwas committed:
another else fix...

qrwas committed:
Merge branch 'master' into Add-eyes-blinking

qrwas committed:
some refactoring

qrwas committed:
add other species

qrwas committed:
try fix blinking for changeling

qrwas committed:
half working changeling

qrwas committed:
cleanup EyeBlinkingSystem

qrwas committed:
Update SharedEyeBlinkingSystem.cs

qrwas committed:
Update EyeBlinkingSystem.cs

qrwas committed:
Merge branch 'master' into Add-eyes-blinking

qrwas committed:
some shitty remade

qrwas committed:
Merge branch 'master' into Add-eyes-blinking

qrwas committed:
change logic to get eyelids colors

qrwas committed:
try add status effect

qrwas committed:
fix blink dyspraxia effect

qrwas committed:
cleanup and add xml documentation

qrwas committed:
fix mouse blinking

qrwas committed:
fix gauzes

qrwas committed:
rename variable

metalgearsloth reviewed (commented):
Code needs a lot of documenting around the systems in particular.

qrwas committed:
add more comments

qrwas committed:
idiot fix

metalgearsloth reviewed (changes requested):
Okay so apparently eyeclosing exists and I really don't want to have to maintain 2 systems so one of them will need to be merged into the other.

qrwas committed:
fix blinking after changeApperanceData and add action for close eye

qrwas committed:
remove unused system and components

qrwas committed:
event handlers name cleanup

qrwas committed:
Merge branch 'master' into Add-eyes-blinking

qrwas committed:
I hate changeling

qrwas committed:
remove not needed additive set EyeClosed to fix failed test

qrwas committed:
Merge branch 'master' into Add-eyes-blinking

qrwas committed:
Merge branch 'master' into Add-eyes-blinking

kontakt committed:
Typos, formatting, subscriptions

Signed-off-by: Connor Huffine <chuffine@gmail.com>

kontakt committed:
[SubscribeNetworkEvent] exists

Signed-off-by: Connor Huffine <chuffine@gmail.com>

kontakt committed:
oops

Signed-off-by: Connor Huffine <chuffine@gmail.com>

kontakt reviewed (changes requested):
More discussion in discord - approach needs to shift to align  with how bodies are supposed to work. Some of the comments remain valid.

qrwas committed:
apply review

qrwas committed:
Start Refactoring EyeBlinkingSystem

qrwas committed:
Remove Client blinking comp

qrwas committed:
add new BodyRelay events

qrwas committed:
Fix sleeping

qrwas committed:
add correct add action

qrwas committed:
changeling...

qrwas committed:
Update SharedEyeBlinkingSystem.cs

qrwas committed:
pupupu....

qrwas committed:
pupupu....

qrwas committed:
Update SharedEyeBlinkingSystem.cs

qrwas committed:
add arachid as new tested entity, and some refactoring init component

qrwas committed:
Try fix some another init

qrwas committed:
Fix changeling transform

qrwas committed:
Update other species

qrwas committed:
Add displacement support

qrwas committed:
fix status effect

qrwas committed:
another fix status effect

qrwas committed:
fix changeling darkest eyelid color after many transformations

qrwas committed:
Update SharedEyeBlinkingSystem.cs

whatston3 commented on Content.Client/Eye/Blinking/EyeBlinkingSystem.cs:
```diff
@@ -0,0 +1,420 @@
+    [Dependency] private StatusEffectsSystem _statusEffects = default!;
+    [Dependency] private IGameTiming _timing = default!;
+    [Dependency] private SharedAppearanceSystem _appearance = default!;
+    [Dependency] private DisplacementMapSystem _displacement = default!;
```
```suggestion
    [Dependency] private IGameTiming _timing = default!;
    [Dependency] private IResourceCache _resCache = default!;
    [Dependency] private IRobustRandom _random = default!;
    [Dependency] private DisplacementMapSystem _displacement = default!;
    [Dependency] private SharedAppearanceSystem _appearance = default!;
    [Dependency] private SpriteSystem _sprite = default!;
    [Dependency] private StatusEffectsSystem _statusEffects = default!;
```
Alphabetize these.

whatston3 committed:
eye blinking suggestions

whatston3 committed:
Blinking fixups

whatston3 committed:
fix action restoration on death

whatston3 reviewed (changes requested):
Also requesting changes - see the above branch! :^)

whatston3 committed:
suggested changes, cleanup & prediction

whatston3 committed:
cloning, action removal

qrwas committed:
Merge pull request #1 from whatston3/2026-08-08-add-eyes-blinking-suggestions

2026 08 08 add eyes blinking suggestions

qrwas committed:
fix diona meta

qrwas committed:
fix incorect target by TryComp

qrwas committed:
add reinit eye blinking if changes rsi path and update logic to update eyelids color for not organ eyes

qrwas committed:
fix remove layers

qrwas committed:
check all layers for correct remove layers

qrwas committed:
Try optimization remove existing layers

qrwas committed:
Update EyeBlinkingSystem.cs

whatston3 reviewed (approved):
A few oddities:
1. Blink dyspraxia doesn't seem to copy over to a clone, not quite sure why.  (Maybe related to #41822?  No idea.)
2. The enabled status on the "close eyes" action is true whenever the mob seems to have its eyes closed.

Only things I could find in testing, but I did leave a few small comments.

All that aside, though, think this looks really good!

qrwas committed:
Update Content.Server/Eye/Blinking/EyeBlinkingSystem.cs

Co-authored-by: Whatstone <166147148+whatston3@users.noreply.github.com>

qrwas committed:
Merge branch 'master' into Add-eyes-blinking

qrwas committed:
apply review

whatston3 commented on Content.Client/Eye/Blinking/EyeBlinkingSystem.cs:
```diff
@@ -0,0 +1,394 @@
+            foreach (var effect in effects)
+            {
+                maxAsyncBlink = maxAsyncBlink.Add(effect.Comp1.MaxAsyncBlink);
+                maxAsyncOpenBlink = maxAsyncOpenBlink.Add(effect.Comp1.MaxAsyncOpenBlink);
```
```suggestion
                maxAsyncBlink += effect.Comp1.MaxAsyncBlink;
                maxAsyncOpenBlink += effect.Comp1.MaxAsyncOpenBlink;
```
terse

whatston3 commented on Content.Server/Eye/Blinking/EyeBlinkingSystem.cs:
```diff
@@ -0,0 +1,44 @@
+            return;
+
+        var ev = new BlinkEyeEvent(GetNetEntity(ent.Owner));
+        RaiseNetworkEvent(ev);
```
Looking at it again, SendEmoteMessage should probably be a `TrySendEmoteMessage` (or `TryEmote` :shrug:) - you should really only set Handled to true if you end up sending the BlinkEyeEvent, no?

whatston3 reviewed (commented):
Zombification doesn't seem to affect eyelid colour at the moment - it likely should, and offhand I don't know how that's handled.
